### DISCUSSION TITLE
[Chapter 05] JPA 엔티티에 대한 행동 설계

### DISCUSSION THREAD

**Discussion by corock** [cowaine/is-my-code-that-strange#39] [Chapter 04-05 (불변, 응집도)]
> 해당 질문은 JPA 사전 지식이 필요할 수 있습니다.

5장에서는 높은 응집도를 나타내기 위해 `GiftPoint`, `Location`, `MagicPoint` 클래스 등을 예시로 들었습니다.
그런데 개발하다 보면 비즈니스에 따라서 해당 클래스가 데이터베이스 테이블에 대응하는 실제 엔티티일 수 있습니다.

해당 책에서는 운 좋게도(?) 이에 해당하는 예시가 구체적으로 나와 있지는 않으나, 예를 들면 `GiftPoint` 같은 포인트 도메인입니다.
만약 JPA 를 사용하는 환경에서 `GiftPoint` 데이터가 어떤 특정한 데이터베이스 스키마에 `gift_points` 테이블 컬럼으로 모델링되어 있고,
`GiftPoint` 가 가지는 행동, 즉 여러 가지 메서드가 정의된다면 결국 엔티티가 비즈니스 로직에 종속될 수 있습니다.

_p. 74_ 에 잠깐 언급되었던 `Product` 나 `Cart` 등의 도메인은 더 필요한 오퍼레이션이 많고, 구현해야 할 메서드가 많을 겁니다.
즉, 엔티티가 너무 많은 책임을 지는 상황이 생길 수 있다는 생각이 드는데 이와 관련해서 다른 분들의 생각이 궁금합니다.

**Comment by firewoody237** [comment]
제가 잘 이해한 것인지는 모르겠으나, 아래 두 가지 중 하나 아닌가요?
1. Entity에 비즈니스 로직을 두지않고 Service 레이어에 둠
2. Entity와 Domain을 분리하고, 비즈니스 로직은 Domain에

**Reply by corock** [reply]
네 아마 나올 수 있는 구성은 언급해 주신 것처럼 주로 두 가지 중 하나일 듯합니다.
이 문제에 대해서 고민한 [게시글](https://velog.io/@kevin_/%EC%97%94%ED%8B%B0%ED%8B%B0%EC%97%90-%EB%B9%84%EC%A6%88%EB%8B%88%EC%8A%A4-%EB%A1%9C%EC%A7%81-%EC%B0%AC%EB%B0%98)이 있어 같이 참고하면 좋을 것 같습니다.
